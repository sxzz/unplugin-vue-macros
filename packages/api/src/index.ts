export { MagicString } from '@vue-macros/common'

// WIP
// export * from './vue'
export * from './ts'
export * from './utils'
