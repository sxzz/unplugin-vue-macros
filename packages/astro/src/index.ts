import { type AstroIntegration } from 'astro'
import { type Options } from 'unplugin-vue-macros'
import VitePlugin from 'unplugin-vue-macros/vite'

const name = 'unplugin-vue-macros'

export default function VueMacrosAstroIntegration(
  userOptions: Options = {}
): AstroIntegration {
  return {
    name,
    hooks: {
      'astro:config:setup': ({ updateConfig }) => {
        updateConfig({
          vitePlugin: [VitePlugin(userOptions)],
        })
      },
    },
  }
}
