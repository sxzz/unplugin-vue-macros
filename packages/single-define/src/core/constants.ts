import { HELPER_PREFIX } from '@vue-macros/common'

export const EMIT_VARIABLE_NAME = `${HELPER_PREFIX}_root_emit`

export const PROPS_VARIABLE_NAME = `${HELPER_PREFIX}_root_props`
