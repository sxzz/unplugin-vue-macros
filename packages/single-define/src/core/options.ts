import { type MagicString } from '@vue-macros/common'
import type { Node } from '@babel/types'

export interface TransformOptions {
  id: string
  code: string
  magicString: MagicString
  offset: number
  setupAst: Node
}
