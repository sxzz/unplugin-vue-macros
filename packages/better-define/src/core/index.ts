import { MagicString, getTransformResult, parseSFC } from '@vue-macros/common'
import { analyzeSFC } from '@vue-macros/api'
import type { TSProps } from '@vue-macros/api'
import type {} from '@babel/types'

export const transformBetterDefine = async (code: string, id: string) => {
  const s = new MagicString(code)
  const sfc = parseSFC(code, id)
  if (sfc.script || !sfc.scriptSetup) return

  const offset = sfc.scriptSetup.loc.start.offset
  const result = await analyzeSFC(s, sfc)
  if (result.props) {
    processProps(result.props)
  }

  return getTransformResult(s, id)

  async function processProps(props: TSProps) {
    const runtimeDefs = await props.getRuntimeDefinitions()

    // TODO prod, default
    const runtimeDecls = `{\n  ${Object.entries(runtimeDefs)
      .map(
        ([key, { type, required }]) =>
          `${key}: { type: ${toRuntimeTypeString(
            type
          )}, required: ${required} }`
      )
      .join(',\n  ')}\n}`

    s.overwriteNode(props.callExpressionAst, `defineProps(${runtimeDecls})`, {
      offset,
    })
  }

  function toRuntimeTypeString(types: string[]) {
    return types.length > 1 ? `[${types.join(', ')}]` : types[0]
  }
}
