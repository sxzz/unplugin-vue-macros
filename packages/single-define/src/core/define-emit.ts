import { DEFINE_EMIT, isCallOf, walkAST } from '@vue-macros/common'

import { type TransformOptions } from './options'

import { EMIT_VARIABLE_NAME } from './constants'
import type { Node } from '@babel/types'

function mountEmits(emits: string[][]) {
  const isAllWithoutOptions = emits.every(([, options]) => !options)

  if (isAllWithoutOptions) {
    return `[${emits.map(([name]) => `'${name}'`).join(', ')}]`
  }

  return `{
      ${emits
        .map(([name, options]) => `${name}: ${options || `{}`}`)
        .join(', ')}
    }`
}

export function transformDefineEmit({
  setupAst,
  code,
  offset,
  magicString,
}: TransformOptions) {
  const emits: string[][] = []

  walkAST<Node>(setupAst, {
    enter(node: Node) {
      if (isCallOf(node, DEFINE_EMIT)) {
        let options = undefined

        const [name, validataion] = node.arguments as any[]

        if (validataion) {
          options = code.slice(
            validataion.start! + offset,
            validataion.end! + offset
          )
        }

        emits.push([name.value, options])

        magicString.overwriteNode(
          node,
          // add space for fixing mapping
          `(payload) => ${EMIT_VARIABLE_NAME}('${name.value}', payload)`,
          { offset }
        )
      }
    },
  })

  if (emits.length > 0) {
    magicString.prependLeft(
      offset!,
      `\n const ${EMIT_VARIABLE_NAME} = defineEmits(${mountEmits(emits)})\n`
    )
  }
}
