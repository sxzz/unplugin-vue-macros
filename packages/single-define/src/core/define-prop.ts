import {
  DEFINE_PROP,
  HELPER_PREFIX,
  importHelperFn,
  isCallOf,
  walkAST,
} from '@vue-macros/common'
import { type TransformOptions } from './options'
import type { Node } from '@babel/types'

function mountProps(props: string[][]) {
  const isAllWithoutOptions = props.every(([, options]) => !options)

  if (isAllWithoutOptions) {
    return `[${props.map(([name]) => `'${name}'`).join(', ')}]`
  }

  return `{
    ${props.map(([name, options]) => `${name}: ${options || `{}`}`).join(', ')}
  }`
}

export const propsVariableName = `${HELPER_PREFIX}_root_props`

export function transformDefineProp({
  setupAst,
  code,
  offset,
  magicString,
}: TransformOptions) {
  const props: string[][] = []

  walkAST<Node>(setupAst, {
    enter(node: Node) {
      if (isCallOf(node, DEFINE_PROP)) {
        let options = undefined

        const [name, definition] = node.arguments as any[]

        if (definition) {
          options = code.slice(
            definition.start! + offset,
            definition.end! + offset
          )
        }

        props.push([name.value, options])

        magicString.overwriteNode(
          node,
          // add space for fixing mapping
          `${HELPER_PREFIX}computed(() => ${propsVariableName}.${name.value})`,
          { offset }
        )
      }
    },
  })

  if (props.length > 0) {
    importHelperFn(magicString, offset, 'computed', 'vue')

    magicString.prependLeft(
      offset!,
      `\n const ${propsVariableName} = defineProps(${mountProps(props)})\n`
    )
  }
}
