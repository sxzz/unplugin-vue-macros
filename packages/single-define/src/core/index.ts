import {
  DEFINE_EMIT,
  DEFINE_PROP,
  MagicString,
  getTransformResult,
  parseSFC,
} from '@vue-macros/common'

import { transformDefineEmit } from './define-emit'
import { transformDefineProp } from './define-prop'
import { type TransformOptions } from './options'

export function transformDefineSingle(code: string, id: string) {
  const { scriptSetup, getSetupAst } = parseSFC(code, id)

  if (!scriptSetup) return

  const offset = scriptSetup.loc.start.offset
  const magicString = new MagicString(code)
  const setupAst = getSetupAst()!

  const options: TransformOptions = {
    code,
    id,
    magicString,
    offset,
    setupAst,
  }

  if (code.includes(DEFINE_EMIT)) {
    transformDefineEmit(options)
  }

  if (code.includes(DEFINE_PROP)) {
    transformDefineProp(options)
  }

  return getTransformResult(magicString, id)
}
