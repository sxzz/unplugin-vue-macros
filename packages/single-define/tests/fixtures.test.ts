import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformDefineSingle } from '../src/core'

describe('fixtures', async () => {
  await testFixtures(
    import.meta.glob('./fixtures/*.{vue,js,ts}', {
      eager: true,
      as: 'raw',
    }),
    (args, id, code) => transformDefineSingle(code, id)?.code
  )
})
