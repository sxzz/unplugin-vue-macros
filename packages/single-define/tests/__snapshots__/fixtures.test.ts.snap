// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > ./fixtures/define-emit.basic.vue 1`] = `
"<script setup lang=\\"ts\\">
 const __MACROS__root_emit = defineEmits(['foo'])

import { expectTypeOf } from 'expect-type'

const foo = (payload) => __MACROS__root_emit('foo', payload)

expectTypeOf(foo).toEqualTypeOf<() => void>()

</script>

<template>
  <div></div>
</template>
"
`;

exports[`fixtures > ./fixtures/define-emit.detalhed.vue 1`] = `
"<script setup lang=\\"ts\\">
 const __MACROS__root_emit = defineEmits({
      foo: (payload) => typeof payload === 'string'
    })

import { expectTypeOf } from 'expect-type'

const foo = (payload) => __MACROS__root_emit('foo', payload)

expectTypeOf(foo).toEqualTypeOf<(payload: string) => void>()

</script>

<template>
  <div></div>
</template>
"
`;

exports[`fixtures > ./fixtures/define-prop.basic.vue 1`] = `
"<script setup lang=\\"ts\\">
 const __MACROS__root_props = defineProps(['foo'])

import { computed as __MACROS_computed } from 'vue';
import { expectTypeOf } from 'expect-type'
import type { ComputedRef } from 'vue'

const foo = __MACROS_computed(() => __MACROS__root_props.foo)

expectTypeOf(foo).toEqualTypeOf<ComputedRef<string>>()

</script>

<template>
  <div></div>
</template>
"
`;

exports[`fixtures > ./fixtures/define-prop.detalhed.vue 1`] = `
"<script setup lang=\\"ts\\">
 const __MACROS__root_props = defineProps({
    foo: {
    type: String,
    required: true,
    default: 'foo'
}, bar: {
    type: [Number, Array],
    required: true,
    default: () => [1, 2, 3]
}
  })

import { computed as __MACROS_computed } from 'vue';
import { expectTypeOf } from 'expect-type'
import type { ComputedRef } from 'vue'

const foo = __MACROS_computed(() => __MACROS__root_props.foo)

const bar = __MACROS_computed(() => __MACROS__root_props.bar)

expectTypeOf(foo).toEqualTypeOf<ComputedRef<string>>()
expectTypeOf(bar).toEqualTypeOf<ComputedRef<number[]>>()

</script>

<template>
  <div></div>
</template>
"
`;
